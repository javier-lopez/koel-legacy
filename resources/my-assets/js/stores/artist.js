/* eslint camelcase: ["error", {properties: "never"}] */

import Vue from 'vue'
import { union, difference, take, orderBy } from 'lodash'

import stub from '@/stubs/artist'

const UNKNOWN_ARTIST_ID = 1
const VARIOUS_ARTISTS_ID = 2

export const artistStore = {
  stub,
  cache: [],

  state: {
    artists: []
  },

  init (artists) {
    this.all = artists

    // Traverse through artists array to get the cover and number of songs for each.
    this.all.forEach(artist => this.setupArtist(artist))
  },

  setupArtist (artist) {
    Vue.set(artist, 'playCount', 0)
    Vue.set(artist, 'info', null)
    Vue.set(artist, 'albums', [])
    Vue.set(artist, 'songs', [])

    this.cache[artist.id] = artist

    return artist
  },

  get all () {
    return this.state.artists
  },

  set all (value) {
    this.state.artists = value
  },

  byId (id) {
    return this.cache[id]
  },

  /**
   * Adds an artist/artists into the current collection.
   *
   * @param  {Array.<Object>|Object} artists
   */
  add (artists) {
    [].concat(artists).forEach(artist => {
      this.setupArtist(artist)
      artist.playCount = artist.songs.reduce((count, song) => count + song.playCount, 0)
    })

    this.all = union(this.all, artists)
  },

  purify () {
    this.compact()
  },

  /**
   * Remove empty artists from the store.
   */
  compact () {
    const emptyArtists = this.all.filter(artist => artist.songs.length === 0)
    if (!emptyArtists.length) {
      return
    }

    this.all = difference(this.all, emptyArtists)
    emptyArtists.forEach(artist => delete this.cache[artist.id])
  },

  isVariousArtists: artist => artist.id === VARIOUS_ARTISTS_ID,

  isUnknownArtist: artist => artist.id === UNKNOWN_ARTIST_ID,

  getSongsByArtist: artist => artist.songs,

  getMostPlayed (n = 6) {
    // Only non-unknown artists with actually play count are applicable.
    // Also, "Various Artists" doesn't count.
    const applicable = this.all.filter(artist => {
      return artist.playCount &&
        !this.isUnknownArtist(artist) &&
        !this.isVariousArtists(artist)
    })

    return take(orderBy(applicable, 'playCount', 'desc'), n)
  }
}
