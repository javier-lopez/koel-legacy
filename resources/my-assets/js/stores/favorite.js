import { difference, union } from 'lodash'
import NProgress from 'nprogress'

import { http } from '@/services'
import { alerts, pluralize } from '@/utils'

export const favoriteStore = {
  state: {
    songs: [],
    length: 0,
    fmtLength: ''
  },

  get all () {
    return this.state.songs
  },

  set all (value) {
    this.state.songs = value
  },

  toggleOne (song) {
    // Don't wait for the HTTP response to update the status, just toggle right away.
    // This may cause a minor problem if the request fails somehow, but do we care?
    song.liked = !song.liked
    song.liked ? this.add(song) : this.remove(song)

    NProgress.start()

    return new Promise((resolve, reject) => {
      http.post('interaction/like', { song: song.id }, ({ data }) => {
        // We don't really need to notify just for one song.
        resolve(data)
      }, error => reject(error))
    })
  },

  /**
   * Add a song/songs into the store.
   *
   * @param {Array.<Object>|Object} songs
   */
  add (songs) {
    this.all = union(this.all, [].concat(songs))
  },

  /**
   * Remove a song/songs from the store.
   *
   * @param {Array.<Object>|Object} songs
   */
  remove (songs) {
    this.all = difference(this.all, [].concat(songs))
  },

  clear () {
    this.all = []
  },

  like (songs) {
    // Don't wait for the HTTP response to update the status, just set them to Liked right away.
    // This may cause a minor problem if the request fails somehow, but do we care?
    songs.forEach(song => { song.liked = true })
    this.add(songs)

    NProgress.start()

    return new Promise((resolve, reject) => {
      http.post('interaction/batch/like', { songs: songs.map(song => song.id) }, ({ data }) => {
        alerts.success(`Added ${pluralize(songs.length, 'song')} into Favorites.`)
        resolve(data)
      }, error => reject(error))
    })
  },

  unlike (songs) {
    songs.forEach(song => { song.liked = false })
    this.remove(songs)

    NProgress.start()

    return new Promise((resolve, reject) => {
      http.post('interaction/batch/unlike', { songs: songs.map(song => song.id) }, ({ data }) => {
        alerts.success(`Removed ${pluralize(songs.length, 'song')} from Favorites.`)
        resolve(data)
      }, error => reject(error))
    })
  }
}
