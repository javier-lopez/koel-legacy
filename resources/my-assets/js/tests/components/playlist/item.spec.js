import Component from '@/components/playlist/item.vue'
import factory from '@/tests/factory'
import { playlistStore } from '@/stores'
import { mock } from '@/tests/__helpers__'

describe('component/playlist/item', () => {
  let playlist
  beforeEach(() => {
    playlist = factory('playlist', {
      id: 99,
      name: 'Foo'
    })
  })

  afterEach(() => {
    jest.resetModules()
    jest.clearAllMocks()
  })

  it('renders a playlist menu item', () => {
    const wrapper = shallow(Component, {
      propsData: { playlist }
    })
    expect(wrapper).toMatchSnapshot()
  })

  it('renders the Favorites menu item', () => {
    const wrapper = shallow(Component, {
      propsData: {
        playlist: {
          name: 'Favorites'
        },
        type: 'favorites'
      }
    })
    expect(wrapper.find('a[href="#!/favorites"]').text()).toMatch('Favorites')
  })

  it('edits a playlist', () => {
    const updateStub = mock(playlistStore, 'update')
    const wrapper = shallow(Component, {
      propsData: { playlist }
    })
    wrapper.dblclick('li.playlist')
    wrapper.blur('input[type=text]')
    expect(updateStub).toHaveBeenCalledWith(playlist)
  })

  it("doesn't allow editing Favorites item", () => {
    const wrapper = shallow(Component, {
      propsData: {
        playlist: { name: 'Favorites' },
        type: 'favorites'
      }
    })
    wrapper.dblclick('li.favorites')
    expect(wrapper.has('input[type=text]')).toBe(false)
  })
})
