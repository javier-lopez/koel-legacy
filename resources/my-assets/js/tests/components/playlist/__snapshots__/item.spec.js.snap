// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component/playlist/item renders a playlist menu item 1`] = `
<li class="playlist playlist"><a href="#!/playlist/99" class="">Foo</a>
  <!---->
</li>
`;
